### DISCUSSION TITLE
inner outer function is supported in OpenModelica?

### DISCUSSION THREAD

**Discussion by mpradovelasco** [OpenModelica/OpenModelica#10665] [Q&A] [answered]
Hello, 

This is my first contribution in this forum. 
I am developing a physiological - cybernetic M&S system based on Modelica (CybSim), which will be delivered as an open system. I would like to use the inner / outer construction for functions, since this one facilitates the addressing of biochemical properties under the CybSim design. 
My tests with that feature have failed (with error messages such as "[The model can't be instantiated due to top-level outer element ‘solventProp‘, it may only be used as part of a simulation). 
This first message tries only to know if the inner / outer function is currently supported in OpenModelica 1.21. Otherwise, is there a dated plannification to add this feature?

Thank you in advance! 
Manuel Prado

**Comment by casella** [comment]
Inner/outer has been supported in omc since the Dark Ages :)

Can you post the model that gives you trouble? As I understand, you are trying to simulate a component that contains an outer reference, but is not contained in another model that has a corresponding inner component.

**Comment by perost** [comment]
It sounds like you're declaring the outer function in the model you're trying to simulate, i.e.:
```
partial function A
  input Real x;
  output Real y;
end A;

function B
  input Real x;
  output Real y = x;
end B;

model MO
  outer function f = A;
  Real x = f(time);
end MO;

model MI
  inner function f = B;
  MO mo;
end MI;
```
If you try to instantiate `MO` you'll get the error since it contains a top-level outer element, so it can neither find a corresponding inner element nor add one automatically since it would cause a conflict with the existing outer element. But instantiating `MI` works as expected since it defines an inner function for the instance of `MO`.

**Comment by mpradovelasco** [comment]
Thank a lot for your agile and fast answers!

I use inner outer structures profusely in the design with success, but not with functions. Effectively, I try to apply the scheme that Perost indicates. I will review again this one and comment here my errors (I hope to find it :)), or post the model with the details. 

Thanks!

**Comment by mpradovelasco** [comment]
Hi again,

I present a simplified version of the code structure (simply removing inheritances and other details that do not affect (I hope). This is:

```
model mechType
	outer function solventProp = PQ.getSolvPropInterface; //1
	....
equation
	var2 = var1*solventProp();
	...
end mechType;

model flt
	inner outer function solventProp = PQ.getSolvPropInterface; //2
	mechType	mech; 
	...
equation
	vsolvent = msolvent*solventProp();
	....
end flt;

model PBPK; 
	inner function solventProp = PQ.getWaterPropRec; //3
	flt tissue1;
	...
equation
	connect (flt1, ...);
	
end PBPK;`
```

The declaration //3 causes the OpenModelica Error cited in the initial doubt (The model can't be instantiated due to top-level outer element ‘solventProp‘, it may only be used as part of a simulation). I have tested a //2 declaration like (removing the inner prefix):
`outer function solventProp = PQ.getSolvPropInterface;`
but the error persists. 

I have worked with other function with input arguments (and even a redeclaration statement to use the function application feauture), and the problem keeps.

I use the same structure with an array (and other Real variables) with success. The reason to define the function in the final model level is that this is the application or user model. There is not problems with the access to PQ package and other variables, since they are used with success. Indeed I can substitute //3 by: 
`inner function solventProp = ANYNAME;`
and the same error persists. 

Certainly, if I remove //3, OM gives the expected warning "An inner declaration for outer class solventProp could not be found and was automatically generated.". I can execute the model correctly (using a different strategy, without this function, for chemical properties), simply removing the //3 declaration. 

Could be that functions do not allow several levels of inner outer? 

Thank you for your help!

**Reply by perost** [reply]
Unfortunately I can't replicate the issue myself, i.e. something like this works fine for me:
```Modelica
partial function A
  input Real x;
  output Real y;
end A;

function B
  input Real x;
  output Real y = x;
end B;

model MO
  outer function f = A;
  Real x = f(time);
end MO;

model MIO
  inner outer function f = B;
  MO mo;
end MIO;

model MI
  inner function f = B;
  MIO mio;
end MI;
```
In this case instantiating MI works fine, while instantiating MIO as expected gives the error you mention.

**Comment by mpradovelasco** [comment]
Thanks again. I have replicate functions A and B and performed several tests in my structure to bound where and when occurs the error. Results are strange. The following structure works fine:

```
partial function A
  input Real x;
  output Real y;
end A;

partial model NullMech
 //outer function f = A;
end mechType;

model NullPBPK
  extends NullMech;
  outer function f = A;   // direct declaration of outer function ...
end NullPBPK;

partial model MasterPBPK
  NullPBPK mech;
end MasterPBPK;

partial model generalPBPK
  extends MasterPBPK;
end generalPBPK;

model flt
  extends generalPBPK;
end flt;

model PBPK;
  funcion B
    input Real x;
    output Real y = x;
  end B;
  inner function f = B;
  flt tissue;
end PBPK;
```
But the following one gives the mentioned error.
```
partial function A
  input Real x;
  output Real y;
end A;

partial model NullMech
  outer function f = A;
end mechType;

model NullPBPK
  extends NullMech;   // inherited declaration of outer function...
end NullPBPK;

partial model MasterPBPK
  NullPBPK mech;
end MasterPBPK;

partial model generalPBPK
  extends MasterPBPK;
end generalPBPK;

model flt
  extends generalPBPK;
end flt;

model PBPK;
  funcion B
   input Real x;
   output Real y = x;
  end B;
  inner function f = B;
  flt tissue;
end PBPK;
```
Codes are equal, excepting that in the second (that gives the error) the outer function declaration is inherited by NullPBPK instead of being declared directly.

**Reply by perost** [reply]
That is indeed a bug, it seems we sometimes failed to find the inner element if the outer element was inherited. I've fixed the issue in #10675, if all goes well it should be included in tomorrows nightly build.

**Reply by mpradovelasco** [reply]
Thank you! I will test it on CybSim once delivered (and I will comment here).

**Reply by perost** [reply, marked as answer]
It seems my fix caused a few regressions on our library testing, so I've improved the fix in #10677 which should hopefully resolve those issues.

**Comment by mpradovelasco** [comment]
Hi again,

I have just used the OM 1.22dev (May 25) to test my CybSim design based on outer/inner function, successfully!

Thank very much for your help and effort!
